### PR TITLE
Closes #5549 Clear cache table on domain change

### LINKED ISSUE #5549: Empty the wpr_rocket_cache table after a domain migration
**Before submitting an issue please check that you’ve completed the following steps:**
yes - Made sure you’re on the latest version
yes - Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**
When a site changes the domain, for example when moving from Staging to Production, the `wpr_rocket_cache` table will carry entries referencing the old domain. Attempting to preload these URLs add extra CPU usage

Deactivating/reactivating the Preload option won't clear these URLs

**To Reproduce**
1. Stagecoach a website with Preload enabled. Doesn't matter if WP Rocket is active or not
2. Disable, Enable the preload, and check the `wpr_rocket_cache` table again. It will contain the old ULRs
3. Start browsing some URLs, these will be added to the cache table
4. See the cache table containing Old and New URLs

**Expected behavior**
After a domain change, the `wpr_rocket_cache` table should be cleared. 

**Screenshots**
TqA8wnY

**Backlog Grooming (for WP Media dev team use only)**
- [ ] Reproduce the problem
- [ ] Identify the root cause
- [ ] Scope a solution
- [ ] Estimate the effort

### PR DESCRIPTION
**Pull request by remyperona**
# Description

Fixes #5549

### User documentation

- Clear the cache table on domain change
- Trigger the OCI warmup on domain change
- Trigger the sitemap preload on domain change

### Technical documentation

- Added a new method to truncate the cache table
- Added a new method to truncate the cache table and start the sitemap preload
- Added a new method callback to trigger the OCI warmup on domain change

## Type of change

- [x] Enhancement (non-breaking change which improves an existing functionality).

# Checklists

## Feature validation

- [x] I validated all the Acceptance Criteria. If possible, provide sreenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Code style
- [x] I wrote self-explanatory code about what it does.
- [x] I wrote comments to explain why it does it.
- [x] I named variables and functions explicitely.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unecessary complexity.

### CODE REVIEW

File: inc/Engine/Preload/Admin/Settings.php [hunk 1]
```diff
@@ -21,15 +22,33 @@ class Settings {
...
-	 * @param PreloadUrl   $preload_url PreloadUrl instance.
+	 * @param Options_Data       $options Instance of options handler.
+	 * @param PreloadUrl         $preload_url PreloadUrl instance.
+	 * @param LoadInitialSitemap $load_initial_sitemap LoadInitialSitemap instance.
+	 * @param CacheTable         $cache_table CacheTable instance.
 	 */
-	public function __construct( Options_Data $options, PreloadUrl $preload_url ) {
-		$this->options     = $options;
-		$this->preload_url = $preload_url;
+	public function __construct( Options_Data $options, PreloadUrl $preload_url, $load_initial_sitemap, $cache_table ) {
```
Comment (jeawhanlee) [inline inc/Engine/Preload/Admin/Settings.php]:
```suggestion
	public function __construct( Options_Data $options, PreloadUrl $preload_url, LoadInitialSitemap $load_initial_sitemap, CacheTable $cache_table ) {
```

File: inc/Engine/Preload/Database/Tables/Cache.php [hunk 2]
```diff
@@ -96,4 +97,17 @@ public function add_is_locked_column() {
 
 		return $this->is_success( $created );
 	}
+
+	/**
+	 * Truncate cache table.
+	 *
+	 * @return bool
+	 */
+	public function truncate_cache_table(): bool {
```
Comment (jeawhanlee) [inline inc/Engine/Preload/Database/Tables/Cache.php:106]: I don't see this method used anywhere in this PR, is it useful?

**Review by jeawhanlee — CHANGES_REQUESTED** [review]

File: inc/Engine/Preload/Database/Tables/Cache.php (hunk 2 above)
Comment (remyperona) [reply inc/Engine/Preload/Database/Tables/Cache.php:106]: Good catch, I updated the function call

**Review by jeawhanlee — APPROVED** [review]

File: inc/Engine/Media/AboveTheFold/WarmUp/Subscriber.php [hunk 3]
```diff
@@ -29,11 +29,12 @@ public function __construct( Controller $controller ) {
...
-			'rocket_after_clear_atf'     => 'warm_up_home',
-			'rocket_saas_api_queued_url' => 'add_wpr_imagedimensions_query_arg',
-			'rocket_job_warmup'          => 'warm_up',
-			'rocket_job_warmup_url'      => 'send_to_saas',
+			'wp_rocket_upgrade'             => [ 'warm_up_on_update', 10, 2 ],
+			'rocket_after_clear_atf'        => 'warm_up_home',
+			'rocket_saas_api_queued_url'    => 'add_wpr_imagedimensions_query_arg',
+			'rocket_job_warmup'             => 'warm_up',
+			'rocket_job_warmup_url'         => 'send_to_saas',
+			'rocket_domain_options_changed' => 'warm_up_home',
```
Comment (jeawhanlee) [inline inc/Engine/Media/AboveTheFold/WarmUp/Subscriber.php]: Something skipped during code review, This event here: `rocket_domain_options_changed` doesn't seem relevant.
We already clear atf [here](https://github.com/wp-media/wp-rocket/blob/e67e7697e368bc5e8bc5d2aeb0c59ffad322c505/inc/Engine/Media/AboveTheFold/Admin/Subscriber.php#L34) during domain and we do this action [here](https://github.com/wp-media/wp-rocket/blob/e67e7697e368bc5e8bc5d2aeb0c59ffad322c505/inc/Engine/Media/AboveTheFold/Admin/Controller.php#L74) when clearing atf which fires the warm up [here](https://github.com/wp-media/wp-rocket/blob/e67e7697e368bc5e8bc5d2aeb0c59ffad322c505/inc/Engine/Media/AboveTheFold/WarmUp/Subscriber.php#L33)

**Review by jeawhanlee — CHANGES_REQUESTED** [review]

**Comment by jeawhanlee** [comment]
Aside the comment above, On develop, I can see that we clear the cache table during domain change but it's not truncated, with this PR we truncate the cache table.

I also proactively ran smoke tests on e2e and it passes. @MathieuLamiot do we wait for a test plan from @wp-media/qa-team  on this one.

**Comment by MathieuLamiot** [comment]
@jeawhanlee @Tabrisrp Unless you see something risky in the changes, let's not wait for a specific test plan. Instead we should:
- Perform the "How to reproduce" steps from the original GH issue and ensure the expected behavior is now happening, that there are no errors/warnings and that the preload feature seems to work appropriately on basic use-cases.
- QA team would then be able to perform overall preload Non-regression tests on develop once they are back.

**Comment by MathieuLamiot** [comment]
Following @jeawhanlee comments on the review, I am moving this back to In Progress so that you can discuss them.

File: inc/Engine/Media/AboveTheFold/WarmUp/Subscriber.php (hunk 3 above)
Comment (remyperona) [reply inc/Engine/Media/AboveTheFold/WarmUp/Subscriber.php]: Good catch

**Review by Miraeld — APPROVED** [review]

**Review by jeawhanlee — APPROVED** [review]

**Comment by jeawhanlee** [comment]
Working as expected according to AC [here](https://github.com/wp-media/wp-rocket/issues/5549#issuecomment-2145981360)

**Pull request merged**